docs/proposals/06-immutable-etcd-backups.md

```diff
@@ -0,0 +1,315 @@
+    - The controller scales in the ETCD cluster (i.e., sets `StatefulSet.spec.replicas` to zero).
+    - The controller creates the `EtcdSnapshotImmutabilityExtension` periodically if `etcd.spec.backup.store.immutableSettings.retentionType` is set to `"Bucket"`.
+
+- **`EtcdSnapshotImmutabilityExtension` specification:**
+  - Run `etcdbrctl extend-immutability --bucket-level-immutability` to extend the immutability of the latest snapshot.
+  - Run `etcdbrctl garbage-collect --garbage-collection-policy <garbage-collection-policy>` to garbage collect the snapshots that are created during the extension.
+
+- **EtcdOperatorTask Controller Logic:**
+  - The operator-tasks controller will react to the creation of the custom resource, and will deploy a job named `<etcd-name>-extend-immutability` which is the `EtcdSnapshotImmutabilityExtension` job.
+  - The controller also reports metrics regarding the `EtcdSnapshotImmutabilityExtension` job, which can be used to raise alerts for operators that immutability has not been extended.
+
+###### Advantages
+
+- **Minimal Operational Impact:** Does not alter the ETCD cluster's state during hibernation and respects the operator's intention to hibernate the cluster without unintended changes.
+- **Efficient Resource Utilization:** Only the latest snapshot is copied, limiting additional storage usage, and avoids the need to start an embedded ETCD instance.
+- **Automated Process:** The process of taking a full snapshot before hibernation and creating the `EtcdSnapshotImmutabilityExtension` is automated within the controller.
+
+###### Disadvantages
+
+- **Additional Complexity:** Requires updates to the etcd controller, introduction of the operator-tasks controller, and introduction of new etcdbrctl commands.
+- **Prerequisite Requirement:** Relies on successfully taking a full snapshot before hibernation, which may introduce delays or require handling snapshot failures.
+
+##### Recommendation
+
+After evaluating both approaches, **Approach 2: Re-upload of the latest snapshot** is recommended due to its minimal operational impact and efficient resource utilization. By ensuring that a full snapshot is taken before hibernation, we maintain data consistency and extend the immutability period effectively. This approach respects the operator's intention to keep the ETCD cluster hibernated without introducing significant resource consumption or complexity.
+
+## Compatibility
+
+The proposed changes are fully compatible with existing ETCD clusters and backup processes.
+
+- **Backward Compatibility:**
+  - Existing clusters without immutable buckets will continue to function without change.
+  - The introduction of the `EtcdSnapshotImmutabilityExtension` does not affect clusters that are not hibernated.
+- **Forward Compatibility:**
```

Comment:
> Can you also talk about any repercussions if immutability is switched on in the bucket, but the relevant settings are not updated in the Etcd spec or etcd-backup-restore processes?

etcd-backup-restore's immutability handling is designed (currently) to be agnostic to the configuration provided to it at startup. It should be able to handle such transient changes during runtime without any issues; as it checks if it can delete a specific object before doing so. If it tries deleting an object and it errors specifically with an error code that corresponds to the attempt to delete an object that is immutable, it handles this error gracefully and skips deletion of that snapshot in that particular garbage collection run.

> For instance, the current garbage collection will start throwing errors.

As explained above, it will not.

> Is etcdbr equipped to handle those?

Yes.

> Should garbage collection skip an object if it is immutable?

It already does.

> I think that might be a prerequisite before turning on immutability on the bucket, don't you think?

As the error handling for immutable snapshots is already present in etcd-backup-restore as of https://github.com/gardener/etcd-backup-restore/releases/tag/v0.32.0, it is safe to switch these settings already.